HPCC-14522 Implement support for affinity in Thor

ghalliday opened the pull request:
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

ghalliday commented:
@jakesmith @mckellyln pushed for discussion.  We also need to make sure it compiles on all our target platforms.

HPCCSmoketest commented:
Automated Smoketest Build: failed
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/smoketest-7984/HPCC-Platform/system/jlib/jthread.cpp:36:18: fatal error: numa.h: No such file or directory
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/jthread.cpp.o] Error 1
make[1]: *** [system/jlib/CMakeFiles/jlib.dir/all] Error 2
make: *** [all] Error 2

ghalliday commented:
@AttilaVamos the smoke test failed because a package numactl-devel would need to be installed.

AttilaVamos commented:
@ghalliday installed.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman can you check if this compiles in osx?

@jakesmith @mckellyln please review.

richardkchapman commented:
It does not - as far as I can determine there is no support for numa or numactl on OSX

ghalliday commented:
ok.  I'll need to possibly make it conditional on numa being available, with a build option.
I'll wait until first review though.

mckellyln commented on cmake_modules/dependencies/el5.cmake:
```diff
@@ -1 +1 @@
-SET_DEPENDENCIES ( CPACK_RPM_PACKAGE_REQUIRES boost141-regex openldap libicu m4 libtool libxslt libxml2 gcc-c++ openssh-server openssh-clients expect libarchive rsync apr apr-util zip tbb )
+SET_DEPENDENCIES ( CPACK_RPM_PACKAGE_REQUIRES boost141-regex openldap libicu m4 libtool libxslt libxml2 gcc-c++ openssh-server openssh-clients expect libarchive rsync apr apr-util zip tbb numactl-devel )
```
Probably dont need -devel package for installation, just numactl.  But do need it for building so its probably fine this way.  But inconsistent with Debian below where its not the -dev package listed.

richardkchapman replied:
These dependencies should list the run-time not compile-time dependencies

ghalliday replied:
I couldn't find a non development package that included the library - if there is one then please let me know what it is.

richardkchapman replied:
http://rpmfind.net/linux/rpm2html/search.php?query=numactl

mckellyln commented on system/jlib/jthread.cpp:
```diff
@@ -2335,3 +2336,104 @@ unsigned threadLogID()  // for use in logging
+        unsigned cpu1 = getCpuId(cur, &next);
+        if (*next == '-')
+        {
+            const char * range = next+1;
```
range is unused and now fails with jlib warnings as errors.

mckellyln commented on system/jlib/jthread.cpp:
```diff
@@ -2335,3 +2336,104 @@ unsigned threadLogID()  // for use in logging
+    if (numNumaNodes <= 1)
+        return;
+
+    unsigned numNodes = 1;
```
numNodes is unused and now fails with jlib warnings as errors

mckellyln commented on system/jlib/jthread.cpp:
```diff
@@ -2335,3 +2336,104 @@ unsigned threadLogID()  // for use in logging
+void bindMemoryToLocalNodes()
+{
+#if defined(LIBNUMA_API_VERSION) && (LIBNUMA_API_VERSION>=2)
+    numa_set_bind_policy(1);
```
Perhaps redundant, but why not also a call here to:
numa_set_localalloc()
just to ensure local node policy, especially if we are enforcing strict.

richardkchapman commented:
@ghalliday  Back to you

mckellyln commented:
For bindMemorytoLocalNodes() we could get the current cpu mask and use that to construct a nodemask and set the mem binding to that, as in for example:

void bindMemoryToLocalNodes()
{
#if defined(LIBNUMA_API_VERSION) && (LIBNUMA_API_VERSION>=2)
    numa_set_bind_policy(1);

    unsigned numNumaNodes = numa_max_node()+1;
    if (numNumaNodes <= 1)
        return;

    struct bitmask *cpus = numa_allocate_cpumask();
    bool ok = (numa_sched_getaffinity(0, cpus) != 0);
    if (ok)
    {
        struct bitmask *nodes = numa_allocate_nodemask();
        int numCpus = numa_num_configured_cpus();
        for (int i=0;i<numCpus;i++)
        {
            bool inSet = numa_bitmask_isbitset(cpus, i);
            if (inSet)
            {
                int node = numa_node_of_cpu(i);
                numa_bitmask_setbit(nodes, node);
            }
        }
        numa_set_membind(nodes);
        DBGLOG("Process memory bound to numa nodemask 0x%x (of %u nodes total)", (unsigned)(*(nodes->maskp)), numNumaNodes);
        numa_bitmask_free(nodes);
    }
    else
        DBGLOG("Failed to get affinity for memory binding");

    numa_bitmask_free(cpus);
#endif
}

mckellyln commented:
Even more simple:

void bindMemoryToLocalNodes()
{
#if defined(LIBNUMA_API_VERSION) && (LIBNUMA_API_VERSION>=2)
    numa_set_bind_policy(1);

    unsigned numNumaNodes = numa_max_node()+1;
    if (numNumaNodes <= 1)
        return;
    struct bitmask *nodes = numa_get_run_node_mask();
    numa_set_membind(nodes);
    DBGLOG("Process memory bound to numa nodemask 0x%x (of %u nodes total)", (unsigned)(*(nodes->maskp)), numNumaNodes);
#endif
}

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@mckellyln Please re-review

ghalliday commented:
I haven't yet looked at mark's memory binding suggestion....  Also we need to think about osx.

richardkchapman commented:
Re OSX, we should just ignore the NUMA calls

richardkchapman commented:
@ghalliday Needs rebasing for a clean merge

ghalliday committed:
HPCC-14522 Implement support for affinity in Thor

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@richardkchapman now rebased.
I have checked it compiles on windows and linux.  Starting a 4 node Thor system now assigns two to socket 0 and two to socket 1 on my local system.

I implemented @mckellyln suggested change for the memory - but that options defaults off because I'm not convinced you want it enabled most of the time.

ghalliday commented:
@richardkchapman

mckellyln commented:
Good to merge
